pkg/querier/api/header_options.go

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/grafana/dskit/middleware"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+
+	//lint:ignore faillint Simple import for the API error package to format the error correctly.
+	apierror "github.com/grafana/mimir/pkg/api/error"
+)
+
+type headerOptionsContextKeyType int
+
+const headerOptionsContextKey headerOptionsContextKeyType = 1
+
+type HeaderOptionsConfig struct {
+	// Headers is a map of Headers to handle and the function to validate the value.
+	// Headers are case insensitive.
+	Headers map[string]func(string) error
+}
+
+type HeaderOptionsContext struct {
+	headers map[string]string
+}
+
+func HeaderOptionsMiddleware(cfg HeaderOptionsConfig) middleware.Interface {
+	return middleware.Func(func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+			hoc := HeaderOptionsContext{headers: make(map[string]string)}
+
+			for header, validate := range cfg.Headers {
+				value := r.Header.Get(header)
+				if value == "" {
+					continue
+				}
+
+				if err := validate(value); err != nil {
+					// Send a Prometheus API-style JSON error response.
+					w.Header().Set("Content-Type", "application/json")
```

Comment:
We have to have at least 3 different versions of functions that already write errors as JSON like this.

Reply:
The one in pkg/api/error [gives me something](https://github.com/grafana/mimir/blob/90bf98d91df3a9404fe0eca37d5e8c2bec0dcb1f/pkg/api/error/error.go#L85) I don't know how to use in the handler. Which one could I reuse?